File: app/src/main/java/com/esorokin/lantector/ui/fragment/detect/DetectLanguageFragment.java
package com.esorokin.lantector.ui.fragment.detect;

import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.BaseTransientBottomBar;
import android.support.design.widget.Snackbar;
import android.support.v7.app.AppCompatActivity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.arellomobile.mvp.presenter.InjectPresenter;
import com.esorokin.lantector.R;
import com.esorokin.lantector.databinding.FragmentDetectLanguageBinding;
import com.esorokin.lantector.presentation.error.UserError;
import com.esorokin.lantector.presentation.presenter.detect.DetectLanguagePresenter;
import com.esorokin.lantector.presentation.view.detect.DetectLanguageView;
import com.esorokin.lantector.ui.fragment.BaseFragment;
import com.esorokin.lantector.ui.plugins.AlertMessagePlugin;
import com.esorokin.lantector.ui.plugins.DialogErrorPlugin;
import com.esorokin.lantector.ui.plugins.ErrorPlugin;
import com.esorokin.lantector.ui.plugins.ProgressPlugin;
import com.esorokin.lantector.ui.plugins.ToolbarPlugin;
import com.esorokin.lantector.utils.ViewUtils;

public class DetectLanguageFragment extends BaseFragment implements DetectLanguageView {
	@InjectPresenter
	DetectLanguagePresenter presenter;

	private FragmentDetectLanguageBinding viewBinding;

	private ErrorPlugin errorPlugin;
	private ProgressPlugin progressPlugin;
	private AlertMessagePlugin alertMessagePlugin;

	@Override
	protected void initPlugins() {
		super.initPlugins();
		compositionPlugin().attach(new ToolbarPlugin((AppCompatActivity) getActivity()));
		compositionPlugin().attach(errorPlugin = new DialogErrorPlugin(this::getContext));
		compositionPlugin().attach(progressPlugin = new ProgressPlugin(this::getContext));
		compositionPlugin().attach(alertMessagePlugin = new AlertMessagePlugin(this::getContext));
	}

	@Nullable
	@Override
	public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
		viewBinding = DataBindingUtil.inflate(inflater, R.layout.fragment_detect_language, container, false);
		return viewBinding.getRoot();
	}

	@Override
	public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
		super.onViewCreated(view, savedInstanceState);

		viewBinding.detectLanguageButton.setOnClickListener(ignore -> {
			ViewUtils.hideKeyboard(getActivity());
			presenter.userClickStartDetect(viewBinding.inputEditText.getText().toString());
		});
	}

	@Override
	public void showDetectingProgress() {
		progressPlugin.showProgress();
	}

	@Override
	public void hideDetectingProgress() {
		progressPlugin.hideProgress();
	}

	@Override
	public void noInternetConnection() {
		Snackbar.make(viewBinding.detectLanguageButton, R.string.error_internet_connection, BaseTransientBottomBar.LENGTH_LONG)
				.setAction(R.string.button_retry, view -> presenter.userClickRetryDetect(viewBinding.inputEditText.getText().toString()))
				.show();
	}

	@Override
	public void showError(UserError error) {
		errorPlugin.showUiError(error, () -> presenter.userHideError());
	}

	@Override
	public void hideError() {
		errorPlugin.hideUiError();
	}

	@Override
	public void showDetectingResult(String language) {
		alertMessagePlugin.showMessage(language.substring(0, 1).toUpperCase() + language.substring(1), () -> presenter.userHideDetectingResult());
	}

	@Override
	public void hideDetectingResult() {
		alertMessagePlugin.hideMessage();
		viewBinding.inputEditText.getText().clear();
	}
}

File: app/src/main/java/com/esorokin/lantector/ui/fragment/detect/DetectLanguageFragment.kt
package com.esorokin.lantector.ui.fragment.detect

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.arellomobile.mvp.presenter.InjectPresenter
import com.esorokin.lantector.R
import com.esorokin.lantector.presentation.error.UserError
import com.esorokin.lantector.presentation.presenter.detect.DetectLanguagePresenter
import com.esorokin.lantector.presentation.view.detect.DetectLanguageView
import com.esorokin.lantector.ui.fragment.BaseFragment
import com.esorokin.lantector.ui.plugin.*
import com.esorokin.lantector.utils.ViewUtils
import com.esorokin.lantector.utils.ext.snack
import kotlinx.android.synthetic.main.fragment_detect_language.*

class DetectLanguageFragment : BaseFragment(), DetectLanguageView {
    @InjectPresenter
    internal lateinit var presenter: DetectLanguagePresenter

    private val errorPlugin: ErrorPlugin = DialogErrorPlugin({ context })
    private val progressPlugin: ProgressPlugin = ProgressPlugin({ activity })
    private val alertMessage: AlertMessagePlugin = AlertMessagePlugin({ activity })
    private val toolbarPlugin: ToolbarPlugin = ToolbarPlugin({ activity as AppCompatActivity }, R.string.screen_title_detect_language)

    override fun initPlugins() {
        super.initPlugins()
        compositionPlugin.attach(errorPlugin)
        compositionPlugin.attach(progressPlugin)
        compositionPlugin.attach(alertMessage)
        compositionPlugin.attach(toolbarPlugin)
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater?.inflate(R.layout.fragment_detect_language, container, false)
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        detectLanguageButton.setOnClickListener {
            ViewUtils.hideKeyboard(activity)
            presenter.userClickStartDetect(inputEditText.text.toString())
        }
    }

    override fun showDetectingProgress() {
        progressPlugin.showProgress()
    }

    override fun hideDetectingProgress() {
        progressPlugin.hideProgress()
    }

    override fun noInternetConnection() {
        detectLanguageButton.snack(getString(R.string.error_internet_connection)) {
            setAction(R.string.button_retry, { presenter.userClickRetryDetect(inputEditText.text.toString()) })
        }
    }

    override fun showError(error: UserError) {
        errorPlugin.showUiError(userError = error, errorHideListener = { presenter.userHideError() })
    }

    override fun hideError() {
        errorPlugin.hideUiError()
    }

    override fun showDetectingResult(language: String) {
        alertMessage.showMessage(
                title = language.capitalize(),
                hideListener = { presenter.userHideDetectingResult() })
    }

    override fun hideDetectingResult() {
        alertMessage.hideMessage()
        inputEditText.text.clear()
    }
}
